File: csharp.NUnit/GildedRose/GildedRose.cs
using System.Collections.Generic;

namespace GildedRoseKata;

// Objetivos:
// - Refatoração da classe GildedRose para encapsular a lógica de atualização da qualidade do item em um método separado.
// - Aprimoramento das regras de atualização da qualidade para itens como "Aged Brie", "Backstage passes" e "Conjured".
// - Adição de testes unitários incluindo "Aged Brie", "Conjured" e "Backstage passes".

public class GildedRose
{
    private IList<Item> Items;

    public GildedRose(IList<Item> Items)
    {
        this.Items = Items;
    }

    public void UpdateQuality()
    {
        for (var i = 0; i < Items.Count; i++)
        {
      UpdateQuality(Items[i]);
        }
    }

    private static void UpdateQuality(Item item) {
            if (item.Name != "Aged Brie" && item.Name != "Backstage passes to a TAFKAL80ETC concert")
            {
                // Se não for Aged Brie ou Backstage passes então executar essa condição
                // Diminui a qualidade em 1 para itens normais
                // Diminui a qualidade em 2 para itens Conjured
                // Não diminui a qualidade para Sulfuras
                if (item.Quality > 0 && item.Name != "Sulfuras, Hand of Ragnaros")
                {
                    // Diminui a qualidade em 1 para itens normais
                    item.Quality--;
                    // Para itens Conjured, diminui mais 1 (total de 2)
                    if (item.Name.Contains("Conjured") && item.Quality > 0)
                    {
                        item.Quality--;
                    }
                }
            }
            else
            {
                if (item.Quality < 50)
                {
                    // Aumenta a qualidade em 1 para Aged Brie e Backstage passes
                    item.Quality++;

                    if (item.Name == "Backstage passes to a TAFKAL80ETC concert")
                    {
                        // Aged Brie e Backstage
                        // A qualidade (quality) aumenta em 2 unidades quando a data de venda (SellIn) é igual ou menor que 10.
                        // A qualidade (quality) aumenta em 3 unidades quando a data de venda (SellIn) é igual ou menor que 5.
                        // A qualidade (quality) do item vai direto à 0 quando a data de venda (SellIn) tiver passado.
                        if (item.SellIn < 11)
                        {
                            if (item.Quality < 50)
                            {
                                item.Quality++;
                            }
                        }

                        if (item.SellIn < 6)
                        {
                            if (item.Quality < 50)
                            {
                                item.Quality++;
                            }
                        }
                    }
                }
            }

            if (item.Name != "Sulfuras, Hand of Ragnaros")
            {
                item.SellIn--;
            }
            // Antes da data de venda expirar
            if (item.SellIn < 0)
            {
                if (item.Name != "Aged Brie")
                {
                    if (item.Name != "Backstage passes to a TAFKAL80ETC concert")
                    {
                        // Depois da data de venda expirar
                        if (item.Quality > 0)
                        {
                            if (item.Name != "Sulfuras, Hand of Ragnaros")
                            {
                                // Diminui a qualidade em 1 para itens normais após a data de venda
                                item.Quality--;
                                // Para itens Conjured, diminui mais 1 (total de 2) após a data de venda
                                if (item.Name.Contains("Conjured") && item.Quality > 0)
                                {
                                    item.Quality--;
                                }
                            }
                        }
                    }
                    else
                    {
                        item.Quality -= item.Quality;
                    }
                }
                else
                {
                    if (item.Quality < 50)
                    {
                        item.Quality++;
                    }
                }
            }
        }
}

File: python/gilded_rose.py
# -*- coding: utf-8 -*-

AGED_BRIE = "Aged Brie"
SULFURAS = "Sulfuras, Hand of Ragnaros"
BACKSTAGE_PASSES = "Backstage passes to a TAFKAL80ETC concert"
CONJURED = "Conjured"
MAX_QUALITY = 50
MIN_QUALITY = 0


class GildedRose(object):

    def __init__(self, items):
        self.items = items

    def update_quality(self):
        for item in self.items:
            self._update_item(item)

    def _update_item(self, item):
        if item.name == SULFURAS:
            return

        if item.name == AGED_BRIE:
            self._increase_quality(item, 1)
        elif item.name == BACKSTAGE_PASSES:
            self._update_backstage_pass_quality(item)
        elif item.name.startswith(CONJURED):
            self._decrease_quality(item, 2)
        else:
            self._decrease_quality(item, 1)

        item.sell_in -= 1

        if item.sell_in < 0:
            self._apply_expired_quality(item)

    def _update_backstage_pass_quality(self, item):
        if item.sell_in <= 5:
            self._increase_quality(item, 3)
        elif item.sell_in <= 10:
            self._increase_quality(item, 2)
        else:
            self._increase_quality(item, 1)

    def _apply_expired_quality(self, item):
        if item.name == AGED_BRIE:
            self._increase_quality(item, 1)
        elif item.name == BACKSTAGE_PASSES:
            item.quality = MIN_QUALITY
        elif item.name.startswith(CONJURED):
            self._decrease_quality(item, 2)
        else:
            self._decrease_quality(item, 1)

    def _increase_quality(self, item, amount):
        item.quality = min(item.quality + amount, MAX_QUALITY)

    def _decrease_quality(self, item, amount):
        item.quality = max(item.quality - amount, MIN_QUALITY)


class Item:
    def __init__(self, name, sell_in, quality):
        self.name = name
        self.sell_in = sell_in
        self.quality = quality

    def __repr__(self):
        return "%s, %s, %s" % (self.name, self.sell_in, self.quality)
